budgetportal/models.py

```diff
@@ -768,8 +769,324 @@ def get_expenditure_over_time(self):
+            virements = {
+                'label': 'Value of virements and shifts due to savings',
+                'amount': int(total_positive_virement_change),
+                'percentage': 100 * float(total_positive_virement_change) / float(total_voted)
+            }
+        return virements if virements else None
+
+    def _get_budget_special_appropriations(self, openspending_api, total_voted):
+        result_for_special_appropriations = openspending_api.aggregate(
+            cuts=[
+                openspending_api.get_financial_year_ref() + ':' + self.get_financial_year().get_starting_year(),
+                openspending_api.get_adjustment_kind_ref() + ':' + '"Special appropriation"',
+
+            ],
+            drilldowns=[
+                openspending_api.get_department_name_ref()
+            ])
+        result_for_special_appropriations = openspending_api.filter_dept(result_for_special_appropriations, self.name)
+
+        total_special_appropriations = 0
+        for cell in result_for_special_appropriations['cells']:
+            if cell['value.sum']:
+                total_special_appropriations += cell['value.sum']
+
+        if total_special_appropriations:
+            return {
+                'amount': total_special_appropriations,
+                'percentage': (float(total_special_appropriations) / float(total_voted)) * 100
+            }
+        else:
+            return None
+
+    def _get_budget_direct_charges(self, openspending_api):
+        result_for_direct_charges = openspending_api.aggregate(
+            cuts=[
+                openspending_api.get_financial_year_ref() + ':' + self.get_financial_year().get_starting_year(),
+                openspending_api.get_programme_name_ref() + ':' + '"Direct charge against the National Revenue Fund"',
+            ],
+            drilldowns=[
+                openspending_api.get_phase_ref(),
+                openspending_api.get_subprogramme_name_ref(),
+                openspending_api.get_department_name_ref(),
+                openspending_api.get_adjustment_kind_ref(),
+            ],
+            order=[openspending_api.get_subprogramme_name_ref()])
+        result_for_direct_charges = openspending_api.filter_dept(result_for_direct_charges, self.name)
+
+        subprog_ref = openspending_api.get_subprogramme_name_ref()
+        phase_ref = openspending_api.get_phase_ref()
+        kind_ref = openspending_api.get_adjustment_kind_ref()
+        subprog_dict = {}
+
+        for cell in result_for_direct_charges['cells']:
+            if cell[kind_ref] == 'Adjustments - Total adjustments':
+                subprog_dict[cell[subprog_ref]] = {
+                    'amount': cell['value.sum'],
+                    'label': cell[subprog_ref]
+                }
+
+        for subprog in subprog_dict.keys():
+            for cell in result_for_direct_charges['cells']:
+                if cell[subprog_ref] == subprog:
```

Comment:
if we write a function like `openspending.filter(cells, list_of_ref_value_tuples) -> matching_cells` we could factor out this kind of code a lot.


It could take an optional argument expected_cell_count which raises an exception if the cell count doesn't match the result.

Reply:
I don't see many repetitions of this type of code in the section we've been working on recently.
Could I look at this again when I need to do more similar operations?